machine-runner/src/runner/runner.ts

```diff
@@ -336,6 +337,7 @@ export const createMachineRunnerInternal = <
               // Effects of handlingReport on emitters
               ;(() => {
                 if (pushEventResult.executionHappened) {
+                  internals.stateChanged = true
```

Comment:
If that's the case, the consequence will be that `reactIntoSelf` not trigger the iteration. Then code like this will have to change.

```
if (state.is(Bidding) && state.bids.length > 3) {
   // do stuff only if there are more than 3 bids
}
```

Reply:
No, when I said “state” I meant all of it, not just the state name. This would fix it, right?

Reply:
Ah, do you mean reference comparison, as in `{} !== {}`?

Reply:
No, I mean value comparison, like `expect(x).toEqual(y)`. The same state can be recomputed due to time travel that has no effect, leading to new object references; emitting a “new” value would be confusing in this case.

Reply:
To use a mechanism similar to `toEqual` would need an assumption on the data inside `payload`, which should be on the scope of the user. Other than that, jest's deep equality mechanism is custom-made, which means we need to make one too.

I see now that time-travel is one concern here. I think it is more accurate to track the last event that triggers the transition, before `caughtUp`. If before and after time travel, the last events of both have different eventId, trigger `state change`

Reply:
Time travel can also result in different events being processed before the final one, so I think we can only judge based on the state and its payload. Isn’t it established practice that deepEquals extends the value equality from primitive types to arrays and objects, recursively? That should be enough, I think.

Reply:
Since we’ll implement our `deepEquals` we can fix the behaviour — e.g. we could also compare Map and Set using value equality. We only need to document it.

Reply:
One more alternative: How about:
- enabling user to write their own equality?
- use deepEqual as default mechanism?

Reply:
We should start without configuration options and only add them when we see that they’re needed.

Reply:
a cursory glance suggests that that implementation should treat Date specifically (but I haven’t reviewed whether correctly) — so we should use that impl?

Reply:
Not sure. It uses `require` which might be harmful to pure es6 code, but I'll try running dev-example for it

Reply:
@rkuhn there's a complication with equality check. 
It is possible that the application code, on a `reactIntoSelf`, simply returns a mutated old state. e.g. 

```
// type { someMap: Map() }

// reaction handler
(ctx) => {
  ctx.self.someMap.delete(someKey)
  return ctx.self
}
```

If we compare the equality this way:
```
const prev = internals.current
// mutation happens here
const next = internals.current
const changed = deepEqual(prev,next)
```

^ This will return true

If we do it this way:
```
const prev = deepCopy(internals.current)
// mutation happens here
const next = internals.current
const changed = deepEqual(prev,next)
```

This is expensive and prone to breakage when there is a new data type. There are data types already non-copyable such as WeakMap, WeakSet. Good `deepCopy` implementation is hard to come by.

Reply:
Let’s not overcomplicate things: we choose one deepEquals that we find reasonable and document that we use exactly that. API users will have to create their state payloads so that they get new states as desired.

Reply:
Wait, I must assert that the complication happens due to the equality check (either deep or not):
- equality checks need 2 values in the first place
- state payload can be non-object

Therefore, either 1.) a copy needs to happen before the comparison, or 2.) the user is told to create their own new value.

> API users will have to create their state payloads so that they get new states as desired.

By this, do you mean option 2?

Reply:
Ah, I misunderstood which point you were talking about: yes, the copy before running the computation is needed, of course. We’ll also need to document that so people know what to expect. I think it is an anti-pattern to have large state payloads, because that implies having large events as well.

Reply:
Changed.
Copy and comparison now only happens on `caughtUp`, not at every computation, which is nice!